include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h

```diff
@@ -136,47 +236,252 @@ struct __parallel_merge_submitter<_IdType, __internal::__optional_kernel_name<_N
+                                                   _Range1ValueType, _Range2ValueType>;
+
+        const std::size_t __n = __rng1.size() + __rng2.size();
+
+        // Empirical number of values to process per work-item
+        const std::uint8_t __chunk = __exec.queue().get_device().is_cpu() ? 128 : 4;
+
+        const _IdType __steps = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __chunk);
+        const _IdType __base_diag_count = 32 * 1'024;
+        const _IdType __steps_between_two_base_diags =
+            oneapi::dpl::__internal::__dpl_ceiling_div(__steps, __base_diag_count);
+
+        return {__base_diag_count, __steps_between_two_base_diags, __chunk, __steps};
+    }
+
+    // Calculation of split points on each base diagonal
+    template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Compare, typename _Storage>
+    sycl::event
+    eval_split_points_for_groups(_ExecutionPolicy&& __exec, _Range1&& __rng1, _Range2&& __rng2, _Compare __comp,
+                                 const nd_range_params& __nd_range_params,
+                                 _Storage& __base_diagonals_sp_global_storage) const
+    {
+        const _IdType __n1 = __rng1.size();
+        const _IdType __n2 = __rng2.size();
+        const _IdType __n = __n1 + __n2;
+
+        return __exec.queue().submit([&__rng1, &__rng2, __comp, __nd_range_params, __base_diagonals_sp_global_storage,
+                                      __n1, __n2, __n](sycl::handler& __cgh) {
+            oneapi::dpl::__ranges::__require_access(__cgh, __rng1, __rng2);
+            auto __base_diagonals_sp_global_acc =
+                __base_diagonals_sp_global_storage.template __get_scratch_acc<sycl::access_mode::write>(
+                    __cgh, __dpl_sycl::__no_init{});
+
+            __cgh.parallel_for<_DiagonalsKernelName...>(
+                sycl::range</*dim=*/1>(__nd_range_params.base_diag_count + 1), [=](sycl::item</*dim=*/1> __item_id) {
+                    auto __global_idx = __item_id.get_linear_id();
+                    auto __base_diagonals_sp_global_ptr =
+                        _Storage::__get_usm_or_buffer_accessor_ptr(__base_diagonals_sp_global_acc);
+
+                    _split_point_t<_IdType> __sp =
+                        __global_idx == 0 ? _split_point_t<_IdType>{0, 0} : _split_point_t<_IdType>{__n1, __n2};
+
+                    if (0 < __global_idx && __global_idx < __nd_range_params.base_diag_count)
+                    {
+                        const _IdType __i_elem =
+                            __global_idx * __nd_range_params.steps_between_two_base_diags * __nd_range_params.chunk;
+                        if (__i_elem < __n)
+                            __sp = __find_start_point(__rng1, __rng2, __i_elem, __n1, __n2, __comp);
+                    }
+
+                    __base_diagonals_sp_global_ptr[__global_idx] = __sp;
+                });
```

Comment:
I would suggest a little bit re-writtig the lambda body to make the code more  readable and understandable at least:

```
{
    auto __global_idx = __item_id.get_linear_id();

    auto __base_diagonals_sp_global_ptr =
        _Storage::__get_usm_or_buffer_accessor_ptr(__base_diagonals_sp_global_acc);

    if(__global_idx == 0)
        __base_diagonals_sp_global_ptr[__global_idx] = split_point_t<_IdType>{0, 0};
    else if (__i_elem < __n)
        __base_diagonals_sp_global_ptr[__global_idx] = __find_start_point(__rng1, __rng2, __i_elem, __n1, __n2, __comp);
    else //__global_idx == __nd_range_params.base_diag_count
        __base_diagonals_sp_global_ptr[__global_idx] = _split_point_t<_IdType>{__n1, __n2};
}
```

Reply:
Will we have any profit in this case?

Reply:
Also I believe your propose isn't equal to the current code logic: you lost the condition `if (__i_elem < __n)`

Reply:
> Will we have any profit in this case? May be performance profit?

As I said "to make the code more readable and understandable at least".

Reply:
if `__global_idx * __base_diag_chunk < __n` is `true`, do we know that `__global_idx < __nd_range_params.base_diag_count` is also `true`?

The two if statements here where one only protects the single multiplication seems fishy to me. If we can consolidate it down to a single if statement then we end up with something pretty similar to @MikeDvorskiy's formulation.  

A multiplication should be less expensive than a branch, so I don't see the need to protect it.  If we need both conditions, I still prefer to consolidate to a single if statement and just use `&&`.

Reply:
Fixed by another way: please take a look.

Reply:
Sure, looks fine.  A ternary version of what Mikhail proposed I think.

Reply:
@MikeDvorskiy hope you like the new variant.